engine/src/em-stack.cpp

```diff
@@ -114,29 +116,79 @@ bool MCStack::view_platform_dirtyviewonresize() const
 	return true;
 }
 
-void
-MCStack::view_platform_updatewindow(MCRegionRef p_dirty_region)
+
+static MCStackUpdateCallback s_updatewindow_callback = nullptr;
+static void *s_updatewindow_context = nullptr;
+
+void MCStack::view_platform_updatewindowwithcallback(MCRegionRef p_region, MCStackUpdateCallback p_callback, void *p_context)
 {
-	/* FIXME implement HiDPI support */
+	s_updatewindow_callback = p_callback;
+	s_updatewindow_context = p_context;
+
+	view_platform_updatewindow(p_region);
+
+	s_updatewindow_callback = nil;
+	s_updatewindow_context = nil;
+}
+
+void MCStack::view_platform_updatewindow(MCRegionRef p_dirty_region)
+{
+	MCRegionRef t_scaled_region;
+	t_scaled_region = nil;
+	
+	MCRegionRef t_screen_region;
+	t_screen_region = nil;
+	
+	MCGFloat t_scale;
+	t_scale = MCResGetPixelScale();
+	
+	if (t_scale != 1.0)
+	{
+		/* UNCHECKED */ MCRegionTransform(p_dirty_region, MCGAffineTransformMakeScale(t_scale, t_scale), t_scaled_region);
+		t_screen_region = t_scaled_region;
+	}
+	else
+		t_screen_region = p_dirty_region;
+	
+	view_device_updatewindow(t_screen_region);
+	
+	if (t_scaled_region != nil)
+		MCRegionDestroy(t_scaled_region);
+}
 
+void MCStack::view_device_updatewindow(MCRegionRef p_region)
```

Comment:
Since most of this handler has been modified - you might as well fix the indentation of the whole thing.

Reply:
Ok, Fixed